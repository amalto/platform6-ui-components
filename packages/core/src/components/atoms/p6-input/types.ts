export enum P6InputType {
  email = 0,
  number = 1,
  password = 2,
  search = 3,
  tel = 4,
  text = 5,
  url = 6,
}
