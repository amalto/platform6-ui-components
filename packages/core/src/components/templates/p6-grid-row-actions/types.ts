import { IconName, IconPrefix } from '@fortawesome/fontawesome-svg-core';
import { EventEmitter } from '@stencil/core';
import { Mode } from '../../../shared/types';
import { RowId } from '../../organisms/p6-grid/core/entities';

export interface ActionsOptions {
  create: { disabled: boolean; canUse: boolean };
  view: { disabled: boolean; canUse: boolean };
  edit: { disabled: boolean; canUse: boolean };
  rename: { disabled: boolean; canUse: boolean };
  delete: { disabled: boolean; canUse: boolean };
  export: { disabled: boolean; canUse: boolean };
  [action: string]: { disabled: boolean; canUse: boolean };
}

export interface Action {
  id: keyof ActionsOptions;
  title: string;
  icon: IconName;
  iconPrefix?: IconPrefix;
  mode?: Mode;
  emitter: () => EventEmitter<{ rowId: RowId }> | undefined;
}
