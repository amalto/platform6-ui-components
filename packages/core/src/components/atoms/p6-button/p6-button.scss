//@import "src/global/_variable";
//@import "src/global/_colors";
//@import "src/global/_functions";
//@import "src/global/_control";

$button-color: $text-strong !default;
$button-background-color: $scheme-main !default;
$button-family: false !default;

$button-border-color: $border !default;
$button-border-width: $control-border-width !default;

$button-padding-vertical: calc(0.5em - #{$button-border-width}) !default;
$button-padding-horizontal: 1em !default;

$button-hover-color: $link-hover !default;
$button-hover-border-color: $link-hover-border !default;

$button-focus-color: $link-focus !default;
$button-focus-border-color: $link-focus-border !default;
$button-focus-box-shadow-size: 0 0 0 0.125em !default;
$button-focus-box-shadow-color: bulmaRgba($link, 0.25) !default;

$button-active-color: $link-active !default;
$button-active-border-color: $link-active-border !default;

$button-text-color: $text !default;
$button-text-decoration: underline !default;
$button-text-hover-background-color: $background !default;
$button-text-hover-color: $text-strong !default;

$button-disabled-background-color: $scheme-main !default;
$button-disabled-border-color: $border !default;
$button-disabled-shadow: none !default;
$button-disabled-opacity: 0.5 !default;

$button-static-color: $text-light !default;
$button-static-background-color: $scheme-main-ter !default;
$button-static-border-color: $border !default;

@import 'src/styles/atoms/control';
@import 'src/styles/atoms/button';

:host {
  display: inline-flex;
  .button {
    @include control;
    @include unselectable;

    background-color: $button-background-color;
    border-color: $button-border-color;
    border-width: $button-border-width;
    color: $button-color;
    cursor: pointer;
    @if $button-family {
      font-family: $button-family;
    }

    justify-content: center;
    padding: $button-padding-vertical $button-padding-horizontal;
    text-align: center;
    white-space: nowrap;

    strong {
      color: inherit;
    }

    ::slotted(.icon:first-child:not(:last-child)),
    ::slotted(p6-icon:first-child:not(:last-child)) {
      margin-left: calc(#{-1 / 2 * $button-padding-horizontal} - #{$button-border-width});
      margin-right: $button-padding-horizontal / 4;
    }

    ::slotted(.icon:last-child:not(:first-child)),
    ::slotted(p6-icon:last-child:not(:first-child)) {
      margin-left: $button-padding-horizontal / 4;
      margin-right: calc(#{-1 / 2 * $button-padding-horizontal} - #{$button-border-width});
    }

    ::slotted(.icon:first-child:last-child),
    ::slotted(p6-icon:first-child:last-child) {
      $margin: calc(#{-1 / 2 * $button-padding-horizontal} - #{$button-border-width});

      margin-left: $margin;
      margin-right: $margin;
    }

    // States
    &:hover,
    &.is-hovered {
      border-color: $button-hover-border-color;
      color: $button-hover-color;
    }

    &:active,
    &.is-active {
      border-color: $button-active-border-color;
      color: $button-active-color;
    }

    &:focus,
    &.is-focused {
      border-color: $button-focus-border-color;
      color: $button-focus-color;
    }

    fieldset[disabled] & {
      background-color: $button-disabled-background-color;
      border-color: $button-disabled-border-color;
      box-shadow: $button-disabled-shadow;
      opacity: $button-disabled-opacity;
    }

    &:focus:not(:active),
    &.is-focused:not(:active) {
      box-shadow: $button-focus-box-shadow-size $button-focus-box-shadow-color;
    }

    // Colors
    &.is-text {
      background-color: transparent;
      border-color: transparent;
      color: $button-text-color;
      text-decoration: $button-text-decoration;

      &:hover,
      &.is-hovered,
      &:focus,
      &.is-focused {
        background-color: $button-text-hover-background-color;
        color: $button-text-hover-color;
      }

      &:active,
      &.is-active {
        background-color: bulmaDarken($button-text-hover-background-color, 5%);
        color: $button-text-hover-color;
      }

      &[disabled],
      fieldset[disabled] & {
        background-color: transparent;
        border-color: transparent;
        box-shadow: none;
      }
    }

    @each $name, $pair in $colors {
      $color: nth($pair, 1);
      $color-invert: nth($pair, 2);
      &.is-#{$name} {
        background-color: $color;
        border-color: transparent;
        color: $color-invert;

        &:hover,
        &.is-hovered {
          background-color: bulmaDarken($color, 2.5%);
          border-color: transparent;
          color: $color-invert;
        }

        &:focus,
        &.is-focused {
          border-color: transparent;
          color: $color-invert;

          &:not(:active) {
            box-shadow: $button-focus-box-shadow-size bulmaRgba($color, 0.25);
          }
        }

        &:active,
        &.is-active {
          background-color: bulmaDarken($color, 5%);
          border-color: transparent;
          color: $color-invert;
        }

        &[disabled],
        fieldset[disabled] & {
          background-color: $color;
          border-color: transparent;
          box-shadow: none;
        }

        &.is-inverted {
          background-color: $color-invert;
          color: $color;

          &:hover,
          &.is-hovered {
            background-color: bulmaDarken($color-invert, 5%);
          }

          &[disabled],
          fieldset[disabled] & {
            background-color: $color-invert;
            border-color: transparent;
            box-shadow: none;
            color: $color;
          }
        }

        &.is-loading {
          &::after {
            border-color: transparent transparent $color-invert $color-invert !important;
          }
        }

        &.is-outlined {
          background-color: transparent;
          border-color: $color;
          color: $color;

          &:hover,
          &.is-hovered,
          &:focus,
          &.is-focused {
            background-color: $color;
            border-color: $color;
            color: $color-invert;
          }

          &.is-loading {
            &::after {
              border-color: transparent transparent $color $color !important;
            }

            &:hover,
            &.is-hovered,
            &:focus,
            &.is-focused {
              &::after {
                border-color: transparent transparent $color-invert $color-invert !important;
              }
            }
          }

          &[disabled],
          fieldset[disabled] & {
            background-color: transparent;
            border-color: $color;
            box-shadow: none;
            color: $color;
          }
        }

        &.is-inverted.is-outlined {
          background-color: transparent;
          border-color: $color-invert;
          color: $color-invert;

          &:hover,
          &.is-hovered,
          &:focus,
          &.is-focused {
            background-color: $color-invert;
            color: $color;
          }

          &.is-loading {
            &:hover,
            &.is-hovered,
            &:focus,
            &.is-focused {
              &::after {
                border-color: transparent transparent $color $color !important;
              }
            }
          }

          &[disabled],
          fieldset[disabled] & {
            background-color: transparent;
            border-color: $color-invert;
            box-shadow: none;
            color: $color-invert;
          }
        }

        // If light and dark colors are provided
        @if length($pair) >= 4 {
          $color-light: nth($pair, 3);
          $color-dark: nth($pair, 4);
          &.is-light {
            background-color: $color-light;
            color: $color-dark;

            &:hover,
            &.is-hovered {
              background-color: bulmaDarken($color-light, 2.5%);
              border-color: transparent;
              color: $color-dark;
            }

            &:active,
            &.is-active {
              background-color: bulmaDarken($color-light, 5%);
              border-color: transparent;
              color: $color-dark;
            }
          }
        }
      }
    }
    // Sizes
    &.is-small {
      @include button-small;
    }

    &.is-normal {
      @include button-normal;
    }

    // Modifiers
    &[disabled] {
      background-color: $button-disabled-background-color;
      border-color: $button-disabled-border-color;
      box-shadow: $button-disabled-shadow;
      opacity: $button-disabled-opacity;
    }

    &.is-fullwidth {
      display: flex;
      width: 100%;
    }

    &.is-loading {
      color: transparent !important;
      pointer-events: none;

      &::after {
        @include loader;
        @include center(1em);

        position: absolute !important;
      }
    }

    &.is-static {
      background-color: $button-static-background-color;
      border-color: $button-static-border-color;
      color: $button-static-color;
      box-shadow: none;
      pointer-events: none;
    }

    &.is-rounded {
      border-radius: $radius-rounded;
      padding-left: calc(#{$button-padding-horizontal} + 0.25em);
      padding-right: calc(#{$button-padding-horizontal} + 0.25em);
    }
  }
}
