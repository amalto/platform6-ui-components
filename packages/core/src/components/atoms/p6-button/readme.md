# p6-button

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute  | Description                                                     | Type                                                                                       | Default        |
| ---------- | ---------- | --------------------------------------------------------------- | ------------------------------------------------------------------------------------------ | -------------- |
| `disabled` | `disabled` | Disabled - If `true`, the user cannot interact with the button. | `boolean`                                                                                  | `false`        |
| `mode`     | `mode`     | set the mode of the button                                      | `Mode.danger \| Mode.default \| Mode.info \| Mode.primary \| Mode.success \| Mode.warning` | `Mode.default` |
| `outlined` | `outlined` | Outlined                                                        | `boolean`                                                                                  | `false`        |
| `size`     | `size`     | set the size of the button                                      | `Size.normal \| Size.small`                                                                | `Size.normal`  |
| `type`     | `type`     | type of the button.                                             | `"button" \| "reset" \| "submit"`                                                          | `'submit'`     |
| `waiting`  | `waiting`  | If set, shows a waiting/busy indicator                          | `boolean`                                                                                  | `false`        |

## Dependencies

### Used by

- [p6-dropdown](../../molecules/p6-dropdown)
- [p6-grid-actions](../../organisms/p6-grid/components/p6-grid-actions)

### Graph

```mermaid
graph TD;
  p6-dropdown --> p6-button
  p6-grid-actions --> p6-button
  style p6-button fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
