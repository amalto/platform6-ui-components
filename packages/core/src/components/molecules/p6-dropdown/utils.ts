export function addDropdownClass(child: Element): Element {
  if (child.nodeName === 'HR') {
    child.classList.add('dropdown-divider');
    return child;
  }

  child.classList.add('dropdown-item');
  return child;
}
