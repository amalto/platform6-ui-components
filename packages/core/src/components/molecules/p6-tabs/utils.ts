export function getTabId(id: string): string {
  return `${id}-tab`;
}

export function isTabValid(tab: Element): tab is HTMLElement {
  return 'title' in tab && 'id' in tab;
}
