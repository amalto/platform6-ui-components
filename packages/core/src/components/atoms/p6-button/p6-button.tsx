import { Component, ComponentInterface, Element, h, JSX, Prop } from '@stencil/core';
import { Mode, Size } from '../../../shared/types';
import { getModeClass, getSizeClass } from '../../../utils/classes';

export type P6ButtonType = 'submit' | 'reset' | 'button';

@Component({
  tag: 'p6-button',
  styleUrl: 'p6-button.scss',
  shadow: true,
})
export class P6Button implements ComponentInterface {
  @Element() host!: HTMLP6ButtonElement;

  /**
   * set the mode of the button
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * Outlined
   */
  @Prop() outlined = false;

  /**
   * If set, shows a waiting/busy indicator
   */
  @Prop() waiting = false;

  /**
   * set the size of the button
   */
  @Prop() size: Size = Size.normal;

  /**
   * type of the button.
   */
  @Prop() type: P6ButtonType = 'submit';

  /**
   * Disabled - If `true`, the user cannot interact with the button.
   */
  @Prop() disabled = false;

  private clickDelegationHandler = (): void => {
    if (this.type === 'submit' || this.type === 'reset') {
      const form = this.host?.closest('form');
      if (form !== null && form !== undefined) {
        const nativeButton = document.createElement('button');
        nativeButton.type = this.type;
        nativeButton.style.display = 'none';
        form.appendChild(nativeButton);
        nativeButton.click();
        form.removeChild(nativeButton);
      }
    }
  };

  render(): JSX.Element {
    const classes = {
      'button': true,
      ...getModeClass(this.mode),
      ...getSizeClass(this.size),
      'is-outlined': this.outlined,
      'is-loading': this.waiting,
    };

    return (
      <button
        class={classes}
        // eslint-disable-next-line react/button-has-type
        type={this.type}
        disabled={this.disabled}
        aria-disabled={this.disabled ? 'true' : null}
        onClick={this.clickDelegationHandler}
      >
        <slot />
      </button>
    );
  }
}
