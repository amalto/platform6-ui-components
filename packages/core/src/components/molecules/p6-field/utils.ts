import { isP6Control, isP6NativeControl, P6Control, P6NativeControl } from '../../../shared/form/control';
import { isInDefaultSlot } from '../../../utils/component';

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function isControl(elmt: unknown): elmt is P6Control<unknown> | P6NativeControl {
  return isP6Control(elmt) || isP6NativeControl(elmt);
}

export function getControl(children: HTMLCollection): P6Control<unknown> | P6NativeControl | undefined {
  return Array.from(children).filter(isInDefaultSlot).find(isControl) as P6Control<unknown> | P6NativeControl | undefined;
}
