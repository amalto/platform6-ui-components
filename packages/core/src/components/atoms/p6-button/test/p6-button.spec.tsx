import { newSpecPage } from '@stencil/core/testing';
import { enumArrayEntryToArray } from '../../../../shared/test/utils';
import { Mode, modes, Size, sizes } from '../../../../shared/types';
import { P6Button } from '../p6-button';

describe('p6-button', () => {
  describe('mode', () => {
    it.each(enumArrayEntryToArray(modes))('has the class "is-%s" when mode is "%s"', async (_, mode) => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button mode=${mode}></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when mode is default', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button mode=${Mode.default}></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when mode is undefined', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button mode=${undefined}></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('outlined', () => {
    it('has a "is-outlined" class when outlined is true', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button outlined="true"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "is-outlined" class when outlined is false', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button outlined="false"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('waiting', () => {
    it('has a "is-loading" class when waiting is true', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button waiting="true"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "is-loading" class when waiting is false', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button waiting="false"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('disabled', () => {
    it('has a "disabled" attribute on the button when disabled is true', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button disabled="true"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('has no "disabled" attribute on the button when disabled is false', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button disabled="false"></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('size', () => {
    it.each(enumArrayEntryToArray(sizes))('has the class "is-%s" when size is "%s"', async (_, size) => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button size=${size}></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('does not add class when size is default', async () => {
      const page = await newSpecPage({
        components: [P6Button],
        html: `<p6-button size=${Size.normal}></p6-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});
