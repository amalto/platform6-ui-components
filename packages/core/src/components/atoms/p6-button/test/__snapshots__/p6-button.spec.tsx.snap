// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-button disabled has a "disabled" attribute on the button when disabled is true 1`] = `
<p6-button disabled="true">
  <mock:shadow-root>
    <button aria-disabled="true" class="button" disabled="" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button disabled has no "disabled" attribute on the button when disabled is false 1`] = `
<p6-button disabled="false">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode does not add class when mode is default 1`] = `
<p6-button mode="2">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode does not add class when mode is undefined 1`] = `
<p6-button mode="undefined">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-danger" when mode is "0" 1`] = `
<p6-button mode="0">
  <mock:shadow-root>
    <button class="button is-danger" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-default" when mode is "2" 1`] = `
<p6-button mode="2">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-info" when mode is "3" 1`] = `
<p6-button mode="3">
  <mock:shadow-root>
    <button class="button is-info" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-primary" when mode is "5" 1`] = `
<p6-button mode="5">
  <mock:shadow-root>
    <button class="button is-primary" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-success" when mode is "4" 1`] = `
<p6-button mode="4">
  <mock:shadow-root>
    <button class="button is-success" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button mode has the class "is-warning" when mode is "1" 1`] = `
<p6-button mode="1">
  <mock:shadow-root>
    <button class="button is-warning" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button outlined has a "is-outlined" class when outlined is true 1`] = `
<p6-button outlined="true">
  <mock:shadow-root>
    <button class="button is-outlined" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button outlined has no "is-outlined" class when outlined is false 1`] = `
<p6-button outlined="false">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button size does not add class when size is default 1`] = `
<p6-button size="1">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button size has the class "is-normal" when size is "1" 1`] = `
<p6-button size="1">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button size has the class "is-small" when size is "0" 1`] = `
<p6-button size="0">
  <mock:shadow-root>
    <button class="button is-small" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button waiting has a "is-loading" class when waiting is true 1`] = `
<p6-button waiting="true">
  <mock:shadow-root>
    <button class="button is-loading" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;

exports[`p6-button waiting has no "is-loading" class when waiting is false 1`] = `
<p6-button waiting="false">
  <mock:shadow-root>
    <button class="button" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
</p6-button>
`;
